src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -194,20 +194,91 @@ public override int GetHashCode()
     /// </summary>
     public class CallGraph
     {
-        // TODO:
-        // This is the method that should be invoked to verify cycles of interest,
-        // i.e. where each callable in the cycle is type parametrized.
-        // It should probably generate diagnostics; I'll add the doc comment once its use is fully defined.
-        internal static bool VerifyCycle(CallGraphNode rootNode, params CallGraphEdge[] edges)
+        /// <summary>
+        /// Given a cycle of call graph edges, determines if the cycle is valid.
+        /// Invalid cycles are those that cause type parameters to be mapped to
+        /// other type parameters of the same callable (constricting resolutions)
+        /// or to a type containing a nested reference to the same type parameter,
+        /// i.e Foo.A -> Foo.A[].
+        /// Returns true if the cycle is valid, false if invalid.
+        /// </summary>
+        internal static bool VerifyCycle(params CallGraphEdge[] edges)
         {
-            var parent = rootNode.CallableName;
             var combination = new TypeResolutionCombination(edges.Select(edge => edge.ParamResolutions).ToArray());
-            var resolvedToConcrete = combination.CombinedResolutionDictionary.FilterByOrigin(parent)
-                .All(kvp => !(kvp.Value.Resolution is ResolvedTypeKind.TypeParameter tp) ||
-                    (tp.Item.Origin.Equals(kvp.Key.Item1) && tp.Item.TypeName.Equals(kvp.Key.Item2)));
-            return combination.IsValid && resolvedToConcrete;
-            //var isClosedCycle = validCycle && combined.Values.Any(res => res.Resolution is ResolvedTypeKind.TypeParameter tp && EqualsParent(tp.Item.Origin));
-            // TODO: check that monomorphization correctly processes closed cycles - meaning add a test...
+            if (!combination.IsValid)
+            {
+                return false;
+            }
+            var allResolutionsValid = combination.CombinedResolutionDictionary
+                .All(kvp => CheckTypeParameterResolutions.IsValidParamResolution(kvp.Key, kvp.Value));
+            return allResolutionsValid;
+        }
+
+        /// <summary>
+        /// Walker that checks a given type parameter resolution to see if it constricts
+        /// the type parameter to another type parameter of the same callable, or contains
+        /// a nested self-reference.
+        /// </summary>
+        private class CheckTypeParameterResolutions : TypeTransformation<CheckTypeParameterResolutions.TransformationState>
+        {
+            private readonly QsQualifiedName origin;
```

Comment:
It seems like it would make more sense to put all fields either in `CheckTypeParameterResolutions` or `TransformationState`, rather than splitting them. If `IsNestedSelfReference` is in `TransformationState` for consistency with multi-class transformations, then `origin` would also need to move into `TransformationState` if the transformation had multiple classes. If not, `IsNestedSelfReference` could also move into a field next to `origin`.

Reply:
Moved to the TransformationState class.